Editor/EditorCore/CutTool.cs

```diff
@@ -0,0 +1,1421 @@
+
+
+namespace UnityEditor.ProBuilder
+{
+    internal class CutTool : EditorTool
+    {
+        ProBuilderMesh m_Mesh;
+
+        Texture2D m_CutCursorTexture;
+        Texture2D m_CutAddCursorTexture;
+        Texture2D m_CutMagnetCursorTexture;
+        Texture2D m_CutAddMagnetCursorTexture;
+        Texture2D m_CurrentCutCursor = null;
+
+        /// <summary>
+        /// Describes the different vertex types on the path.
+        /// </summary>
+        [System.Flags]
+        public enum VertexTypes
+        {
+            None = 0 << 0,
+            NewVertex = 1 << 0,
+            AddedOnEdge = 1 << 1,
+            ExistingVertex = 1 << 2,
+            VertexInShape = 1 << 3,
+        }
+
+        [Serializable]
+        internal struct CutVertexData
+        {
+            public Vector3 m_Position;
+            public Vector3 m_Normal;
+            public VertexTypes m_Types;
+
+            public Vector3 position
+            {
+                get => m_Position;
+                set => m_Position = value;
+            }
+
+            public Vector3 normal
+            {
+                get => m_Normal;
+            }
+
+            public VertexTypes types
+            {
+                get => m_Types;
+            }
+
+            public CutVertexData(Vector3 position, Vector3 normal, VertexTypes types = VertexTypes.None)
+            {
+                m_Position = position;
+                m_Normal = normal;
+                m_Types = types;
+            }
+        }
+
+        static readonly Color k_HandleColor = new Color(.8f, .8f, .8f, 1f);
+        static readonly Color k_HandleColorAddNewVertex = new Color(.01f, .9f, .3f, 1f);
+        static Color s_HandleColorAddVertexOnEdge = new Color(.3f, .01f, .9f, 1f);
+        static Color s_HandleColorUseExistingVertex = new Color(.01f, .5f, 1f, 1f);
+        static readonly Color k_HandleColorModifyVertex = new Color(1f, .75f, .0f, 1f);
+        const float k_HandleSize = .05f;
+
+        static readonly Color k_LineColor = new Color(0f, 55f / 255f, 1f, 1f);
+        static readonly Color k_InvalidLineColor = Color.red;
+        static readonly Color k_ConnectionsLineColor = new Color(0f, 200f / 255f, 170f / 200f, 1f);
+        static readonly Color k_DrawingLineColor = new Color(0.01f, .9f, 0.3f, 1f);
+
+        Color m_CurrentHandleColor = k_HandleColor;
+
+        //Handles and point placement
+        int m_ControlId;
+        bool m_PlacingPoint;
+        internal bool m_SnappingPoint;
+        bool m_ModifyingPoint; // State machine instead? //Status
+        int m_SelectedIndex = -2;
+        bool m_IsCutValid = true;
+
+        //Cut tool elements
+        internal Face m_TargetFace;
+        internal Face m_CurrentFace;
+        internal Vector3 m_CurrentPosition = Vector3.positiveInfinity;
+        internal Vector3 m_CurrentPositionNormal = Vector3.up;
+        internal VertexTypes m_CurrentVertexTypes = VertexTypes.None;
```

Comment:
why are these all internal? who is accessing these outside of the class?

Reply:
The test class CutToolTest.cs is :) (Yeah, I did try to make a test class for this tool :D )